src/sinks/datadog/logs.rs

```diff
@@ -2,59 +2,123 @@ use crate::{
+            BatchConfig, BatchSettings, BoxedRawValue, Compression, Encoding, JsonArrayBuffer,
+            TowerRequestConfig, VecBuffer,
+        },
         Healthcheck, VectorSink,
     },
-    tls::{MaybeTlsSettings, TlsConfig},
 };
 use bytes::Bytes;
-use futures01::{stream::iter_ok, Sink};
+use flate2::write::GzEncoder;
+use futures::FutureExt;
+use futures01::Sink;
+use http::{Request, StatusCode};
+use hyper::body::Body;
 use serde::{Deserialize, Serialize};
+use serde_json::json;
+use std::io::Write;
 
-#[derive(Deserialize, Serialize, Debug)]
+#[derive(Deserialize, Serialize, Debug, Clone)]
 #[serde(deny_unknown_fields)]
 pub struct DatadogLogsConfig {
-    endpoint: Option<UriSerde>,
+    endpoint: Option<String>,
     api_key: String,
     encoding: EncodingConfig<Encoding>,
-    tls: Option<TlsConfig>,
+
+    #[serde(default)]
+    compression: Compression,
+
+    compression_level: Option<u32>,
+
+    #[serde(default)]
+    batch: BatchConfig,
+
+    #[serde(default)]
+    request: TowerRequestConfig,
+}
+
+#[derive(Clone)]
+pub struct DatadogLogsJsonService {
+    config: DatadogLogsConfig,
+}
+
+#[derive(Clone)]
+pub struct DatadogLogsTextService {
+    config: DatadogLogsConfig,
 }
 
 inventory::submit! {
     SinkDescription::new_without_default::<DatadogLogsConfig>("datadog_logs")
 }
 
+impl DatadogLogsConfig {
+    fn get_endpoint(&self) -> &str {
+        match &self.endpoint {
+            Some(ref endpoint) => &endpoint,
+            None => "https://http-intake.logs.datadoghq.eu/v1/input",
+        }
+    }
+}
+
 #[typetag::serde(name = "datadog_logs")]
 impl SinkConfig for DatadogLogsConfig {
     fn build(&self, cx: SinkContext) -> crate::Result<(VectorSink, Healthcheck)> {
-        let (host, port) = if let Some(uri) = &self.endpoint {
-            let host = uri
-                .host()
-                .ok_or_else(|| "A host is required for endpoint".to_string())?;
-            let port = uri
-                .port_u16()
-                .ok_or_else(|| "A port is required for endpoint".to_string())?;
-
-            (host.to_string(), port)
-        } else {
-            ("intake.logs.datadoghq.com".to_string(), 10516)
-        };
+        let request_settings = self.request.unwrap_with(&TowerRequestConfig::default());
+        let client = HttpClient::new(cx.resolver(), None)?;
 
-        let tls_settings = MaybeTlsSettings::from_config(
-            &Some(self.tls.clone().unwrap_or_else(TlsConfig::enabled)),
-            false,
-        )?;
+        // Create a different sink depending on which encoding we have chosen.
+        // Json and text have different batching strategies and so each needs to be
+        // handled differently.
+        match self.encoding.codec {
+            Encoding::Json => {
+                let batch_settings = BatchSettings::default()
+                    .bytes(bytesize::kib(100u64))
+                    .timeout(1)
+                    .parse_config(self.batch)?;
```

Comment:
This appears to be the same in both branches. I suggest just unwrapping before the match.

Reply:
They look the same, but they infer to different types for each branch. For one `batchsettings` is `Batch<VecBuffer<_>>` and for the other branch it is  `Batch<JsonArrayBuffer>`.  It doesn't work before the match as it doesn't know which object to create (believe me, I tried!).

What I can do is move that out into a separate generic function so at least we don't have duplicate code. It's not ideal, if you can think of a better way, let me know!